docs/spec/identities.md

```diff
@@ -0,0 +1,438 @@
+according to the trust relationships between peers ("tracking").
+
+Our model is based on The Update Framework (TUF)[@tuf], conveived as a means of
+securely distributing software packages.
+
+## Data Model
+
+The identity of a repository is established using a document of the form:
+
+```rust
+struct Doc<T, D> {
+    replaces: Option<Revision>,
+    payload: T,
+    delegations: D,
+}
+```
+
+where:
+
+* `Revision` is a cryptographic hash of a document's contents, such that this
+  document is content-addressable by this hash within the storage system.
+* `replaces` refers to the previous revision of the document, or none if it is
+  the first revision
+* `payload` is an extensible, forwards- and backwards-compatible datatype
+  containing application-defined metadata about the repository. The protocol
+  interprets _some_ of the properties, as described in [TODO].
+* `delegations` contains the public keys of key owners who are authorised to
+  issue and approve new revisions of the document. The delegation format depends
+  on the type of identity being established, as detailed below.
+
+The `Doc` MUST be serialised in canonical form, e.g. [Canonical
+JSON](https://wiki.laptop.org/go/Canonical JSON).
+
+The authenticity of the `Doc` is captured by the following type:
+
+```rust
+struct Identity<T, D> {
+    id: ContentId,
+    root: Revision,
+    revision: Revision,
+    doc: Doc<T, D>,
+    signatures: HashMap<PublicKey, Signature>,
+}
+```
+
+where:
+
+* `id` is the content-addressable hash of the `Identity` itself
+* `root` is the initial revision of the identity `Doc`
+* `revision` is the current revision of the identity `Doc`
+* `doc` is the `Doc` described by `revision`
+* `signatures` contains signatures over the `revision`, indexed by the public
+  keys used
+
+An `Identity` describes the attestation of a `Doc`s validity. Attestation chains
+MUST be stored as a hash-linked history (each `Identity` refers to its parent)
+-- we omit the parent `id` as a field here, as it is not used in the following
+sections.
+
+The `root` of a verified `Identity` is the stable identifier of the repository.
+
+> **TODO** Insert specification of URNs here
+
+## Verification
+
+Document revisions, as well as authenticity attestations form a hash-linked
+chain leading to the initial revision of the document. In order to verify the
+authenticity of a given identity attestation (`Identity`), the chain of updates
+must be contiguous and subject to the following verification rules.
+
+We can distinguish four levels of validity:
+
+1. **Untrusted**
+
+    The identity document is well-formed, and points to a root object which is
+    retrievable from local storage.
+
+2. **Signed**
+
+    The identity carrier passes 1., and is signed by at least one key specified
+    in the delegations of the document.
+
+3. **Quorum**
+
+   The identity carrier passes 2., and is signed by a quorum of the keys
```

Comment:
If we require just a quorum of signatures, how do we attest the fact that the user owns the newly introduced keys?
(or, in the case of a project, how do we attest that new maintainers do agree to being maintainers if they do not sign the identity)

Probably the rule should require:

- all signatures if the set of delegations changed
- a quorum of signatures if the set of delegations did not change

Reply:
I don't think this matters -- if a quorum agrees that they trust the new key, the key owner doesn't have to agree to that.

Reply:
Fair enough, this attests _trust_ of a key and not _ownership_ of a key, but it should suffice :smiley: